### PR TITLE
add test targets in makefiles to run local tests

### PR DESCRIPTION
**Pull request by michelleN**

### CODE REVIEW

**Comment by jackgr** [comment]
go test -v ./... can be run from the root, or from any of the folders in the tree. Why add it to make?

**Comment by sgoings** [comment]
This test target looks like it already exists: https://github.com/kubernetes/deployment-manager/blob/master/Makefile#L34

**Comment by jackgr** [comment]
Yes, but only at the top level. Same is true for build. The go tools knows how to recurse, and there's nothing to maintain to make it to the right thing. That said, if we have some other kinds of tests to run, then it would make sense to add a test target to the Makefile for the project(s) containing those tests, and to make the test target recursively descend.

File: Makefile [hunk 1]
```diff
@@ -48,3 +48,6 @@ $(TARGETS) : % : $(addsuffix %,$(SUBDIRS))
 
 $(SUBDIRS_TARGETS) :
 	$(MAKE) -C $(@D) $(@F:.%=%)
+
+test:
+	go test -v ./...
```
Comment (sgoings) [inline Makefile]: @michelleN this is going to cause an issue due to the fact that it's already defined above.

**Comment by sgoings** [comment]
@jackgr if you're following along at home, I just pushed that commit after some time talking to @michelleN about how we could improve the process of [linting, vetting, unit testing, integration testing, ...] in each subproject. @michelleN will be working on it still from here.

**Comment by michelleN** [comment]
This PR should satisfy issue #211. @technosophos: Is this what you were thinking?

File: .travis.yml [hunk 2]
```diff
@@ -9,4 +9,5 @@ install:
   - sudo pip install -r expandybird/requirements.txt
 
 script:
+  - make setup-gotools
   - make test
```
Comment (sgoings) [inline .travis.yml]: You could combine these two lines: `make setup-gotools test`

File: expandybird/Makefile [hunk 3]
```diff
@@ -47,3 +49,21 @@ clean:
...
+test-unit:
+	$(GO_TEST)
+
+.PHONY: lint
+lint:
+	$(GO_LINT)
+
+.PHONY: vet
+vet:
+	$(GO_VET)
```
Comment (sgoings) [inline expandybird/Makefile]: These commands actually could all live in include.mk for less duplication. :smile:

File: include.mk [hunk 4]
```diff
@@ -0,0 +1,4 @@
+SETUP_GOTOOLS := @echo "Installing golint and vet" && go get -u github.com/golang/lint/golint && go get -u -v golang.org/x/tools/cmd/vet
+GO_VET := @echo "Running go vet ..." && go vet ./...
+GO_LINT := @echo "Running golint ..." && golint ./...
+GO_TEST := @echo "Running tests ..." && go test -v ./...
```
Comment (sgoings) [inline include.mk]: If you move the make targets [test, test-unit, lint, vet, setup-gotools] into this file, you'll probably no longer have a need to create make targets inside of make variables.

File: .travis.yml (hunk 2 above)
Comment (michelleN) [inline .travis.yml]: yep. thought about that too. will do :+1:

**Comment by sgoings** [comment]
This looks lots better. Any reason why you didn't put the "test" target in the include.mk file so that it could just be inherited by all projects (sub or root)?

@jackgr I think this boosts deployment-manager contributor and maintainer UX - if someone just works on expandybird they're able to test, vet, lint, etc. from within the expandybird directory just like they would at the root or in another subproject. @michelleN also removed potential duplication (which is what you stated you were worried about), created a model for controlling build complexity down the road, and provided a consistent interface (namely: the test target which runs more than just unit tests now) for external contributors to rely on when they go to work on deployment-manager and internal developers to use even if we decide to tweak how certain subprojects run tests (or potentially get more advanced functional tests). What do you think?

File: manager/Makefile [hunk 5]
```diff
@@ -12,6 +12,8 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+include ../include.mk
+
 .PHONY : all build test push container clean .project
```
Comment (sgoings) [inline manager/Makefile]: This "test" PHONY target is no longer necessary as it's defined near the target below.

File: resourcifier/Makefile [hunk 6]
```diff
@@ -14,6 +14,8 @@
 
 # If you update this image please check the tag value before pushing.
 
+include ../include.mk
+
 .PHONY : all build test push container clean
```
Comment (sgoings) [inline resourcifier/Makefile]: See related comment in manager/Makefile re: duplicate `PHONY: test` statements

File: .travis.yml [hunk 7]
```diff
@@ -9,4 +9,4 @@ install:
   - sudo pip install -r expandybird/requirements.txt
 
 script:
-  - make test
+  - make setup-gotools test
```
Comment (sgoings) [inline .travis.yml]: :+1:

**Comment by sgoings** [comment]
Ooh ooh! Pretty please move the "info" target into the `include.mk` too?

**Comment by jackgr** [comment]
Travis doesn't like it...

`The command "make setup-gotools test" exited with 2.`

**Comment by michelleN** [comment]
@jackgr Fixed some issues raised by go vet. We should be in business now. What do you think?

**Comment by jackgr** [comment]
Nice work, @michelleN!

**Comment by jackgr** [comment]
Fixes #211.

**Pull request merged**